### PR TITLE
Case 21638: RC79 HERO: Adding Hero support in simulation of avatars

### PR DESCRIPTION
**Pull request by samcake**
https://highfidelity.manuscript.com/f/cases/21638/Add-the-Hero-avatars-simulation-prioritization-on-the-client-side

This pr introduces the "hasPriority" flag to the AvatarData
The hasPriority flag is set from the AvatarMixer (never read from the data sent TO the mixer) based on the position  relative to a "hero" zone.

On the client engine side, we do the same logic as in the Mixer, during simulation of OtherAvatars we make 2 queues of avatars to simulate, one for Heroes, one for the rest of the crowd avatars.
We process first the Heroes with a budget of 80% of the OtherAvatars budget, then the rest of the crowd in the remaining time.
THe client just read the hasPriority status of the avatars, it doesn;t evaluate it.

## TEST PLAN

Use this script : https://gist.githubusercontent.com/samcake/40769fe82ea1eae54ccfe9ab61a54b34/raw/e7e5e2d6f064534f428c70d04ada92a26ccc988b/debugAvatarCrowd.js
and visualize the simulation rate vs the network rate.
the color of the indicator per avatar is green at simulation of 30Hz, red at 0hz

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17961/HighFidelity-Beta-PR15129-1ee539c.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17961/HighFidelity-Quest-Beta-PR15129-1ee539c.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17961/HighFidelity-Beta-PR15129-1ee539c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17961/HighFidelity-Beta-Interface-PR15129-1ee539c.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17966/HighFidelity-Beta-PR15129-b87ad32.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17966/HighFidelity-Quest-Beta-PR15129-b87ad32.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17966/HighFidelity-Beta-PR15129-b87ad32.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17966/HighFidelity-Beta-Interface-PR15129-b87ad32.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17966/HighFidelity-Beta-PR15129-b87ad32.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17966/HighFidelity-Beta-Interface-PR15129-b87ad32.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17975/HighFidelity-Beta-PR15129-25814db.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17975/HighFidelity-Quest-Beta-PR15129-25814db.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17975/HighFidelity-Beta-PR15129-25814db.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17975/HighFidelity-Beta-Interface-PR15129-25814db.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17975/HighFidelity-Beta-PR15129-25814db.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17975/HighFidelity-Beta-Interface-PR15129-25814db.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17982/HighFidelity-Beta-PR15129-e38dfee.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17982/HighFidelity-Quest-Beta-PR15129-e38dfee.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17982/HighFidelity-Beta-PR15129-e38dfee.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17982/HighFidelity-Beta-Interface-PR15129-e38dfee.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17982/HighFidelity-Beta-PR15129-e38dfee.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17982/HighFidelity-Beta-Interface-PR15129-e38dfee.exe

File: libraries/avatars/src/AvatarData.h [hunk 1]
```diff
@@ -1121,6 +1125,10 @@ class AvatarData : public QObject, public SpatiallyNestable {
     int getAverageBytesReceivedPerSecond() const;
     int getReceiveRate() const;
 
+    // An Avatar can be set Priority from the AvatarMixer side.
+    bool getHasPriority() const { return _hasPriority; }
```
Comment (SimonWalton-HiFi) [inline libraries/avatars/src/AvatarData.h]: We might want to make reading a scriptable Q_PROPERTY.

**Comment by SimonWalton-HiFi** [comment]
The avatar-protocol parts of this look good to me.

File: libraries/avatars/src/AvatarData.h (hunk 1 above)
Comment (samcake) [reply libraries/avatars/src/AvatarData.h]: not in this version

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17996/HighFidelity-Beta-PR15129-3d11565.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17996/HighFidelity-Quest-Beta-PR15129-3d11565.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17996/HighFidelity-Beta-PR15129-3d11565.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17996/HighFidelity-Beta-Interface-PR15129-3d11565.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17996/HighFidelity-Beta-PR15129-3d11565.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17996/HighFidelity-Beta-Interface-PR15129-3d11565.exe

File: assignment-client/src/avatars/AvatarMixerClientData.cpp [hunk 2]
```diff
@@ -153,6 +154,9 @@ int AvatarMixerClientData::parseData(ReceivedMessage& message, const SlaveShared
         //    qCWarning(avatars) << "Avatar" << _avatar->getSessionDisplayName() << "in hero zone";
         //}
 #endif
+    } else {
+        // if not moved, then ignore the client knwon state (always off) and restore the priority as we know it.
```
Comment (howard-stearns) [inline assignment-client/src/avatars/AvatarMixerClientData.cpp]: as long as we're looking carefully, "knwon" => "known"

File: assignment-client/src/avatars/MixerAvatar.h [hunk 3]
```diff
@@ -19,11 +19,11 @@
 
 class MixerAvatar : public AvatarData {
 public:
-    bool getHasPriority() const { return  _hasPriority; }
-    void setHasPriority(bool hasPriority) { _hasPriority = hasPriority; }
+//    bool getHasPriority() const { return  _hasPriority; }
```
Comment (howard-stearns) [inline assignment-client/src/avatars/MixerAvatar.h]: ??

Comment (SimonWalton-HiFi) [reply assignment-client/src/avatars/MixerAvatar.h]: MixerAvatar is a avatar-mixer-only derived class; now that priority status is moved to the base shared class it doesn't actually do anything.

File: interface/src/avatar/OtherAvatar.cpp [hunk 4]
```diff
@@ -197,7 +197,7 @@ void OtherAvatar::resetDetailedMotionStates() {
 
 void OtherAvatar::setWorkloadRegion(uint8_t region) {
     _workloadRegion = region;
-    QString printRegion = "";
+  /*  QString printRegion = "";
```
Comment (howard-stearns) [inline interface/src/avatar/OtherAvatar.cpp]: kill it?

File: interface/src/avatar/OtherAvatar.cpp [hunk 5]
```diff
@@ -232,7 +232,7 @@ void OtherAvatar::computeShapeLOD() {
     if (newLOD != _bodyLOD) {
         _bodyLOD = newLOD;
         if (isInPhysicsSimulation()) {
-            qCDebug(avatars) << "Changing to body LOD " << newLOD;
+            //qCDebug(avatars) << "Changing to body LOD " << newLOD;
```
Comment (howard-stearns) [inline interface/src/avatar/OtherAvatar.cpp]: kill it

Comment (samcake) [reply interface/src/avatar/OtherAvatar.cpp]: yes

File: interface/src/avatar/OtherAvatar.cpp (hunk 4 above)
Comment (hyperlogic) [inline interface/src/avatar/OtherAvatar.cpp]: Why is this in the PR?

File: interface/src/avatar/OtherAvatar.cpp (hunk 5 above)
Comment (hyperlogic) [inline interface/src/avatar/OtherAvatar.cpp]: Why is this in the PR?

File: libraries/avatars/src/AvatarData.cpp [hunk 6]
```diff
@@ -341,7 +341,7 @@ QByteArray AvatarData::toByteArray(AvatarDataDetail dataDetail, quint64 lastSent
             hasLookAtPosition = sendAll || lookAtPositionChangedSince(lastSentTime);
             hasAudioLoudness = sendAll || audioLoudnessChangedSince(lastSentTime);
             hasSensorToWorldMatrix = sendAll || sensorToWorldMatrixChangedSince(lastSentTime);
-            hasAdditionalFlags = sendAll || additionalFlagsChangedSince(lastSentTime);
+            hasAdditionalFlags = sendAll || additionalFlagsChangedSince(lastSentTime) || getHasPriority();
```
Comment (hyperlogic) [inline libraries/avatars/src/AvatarData.cpp]: I'd be concerned that when an avatar flips from hero to non-hero.  That clients will not be informed.  Leading them to have more hero avatars then strictly necessary.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18001/HighFidelity-Beta-PR15129-99ebff5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18001/HighFidelity-Beta-Interface-PR15129-99ebff5.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18001/HighFidelity-Beta-PR15129-99ebff5.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18001/HighFidelity-Quest-Beta-PR15129-99ebff5.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18001/HighFidelity-Beta-PR15129-99ebff5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18001/HighFidelity-Beta-Interface-PR15129-99ebff5.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18009/HighFidelity-Beta-PR15129-2b6025f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18009/HighFidelity-Quest-Beta-PR15129-2b6025f.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18009/HighFidelity-Beta-PR15129-2b6025f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18009/HighFidelity-Beta-Interface-PR15129-2b6025f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18009/HighFidelity-Beta-PR15129-2b6025f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18009/HighFidelity-Beta-Interface-PR15129-2b6025f.exe

**Pull request merged**